idaes/apps/grid_integration/bidder.py

```diff
@@ -0,0 +1,850 @@
+
+        # populate scenario blocks
+        self.model.fs = pyo.Block(self.model.SCENARIOS)
+        for i in self.model.SCENARIOS:
+            self.bidding_model_object.populate_model(self.model.fs[i])
+
+        # save power output variable in the model object
+        self._save_power_outputs()
+
+        self.formulate_bidding_problem()
+
+        # declare a list to store results
+        self.bids_result_list = []
+
+    def _save_power_outputs(self):
+
+        """
+        Create references of the power output variable in each price scenario
+        block.
+
+        Arguments:
+            None
+
+        Returns:
+            None
+        """
+
+        for i in self.model.SCENARIOS:
+            # get the power output
+            power_output_name = self.bidding_model_object.power_output
+            self.model.fs[i].power_output_ref = pyo.Reference(
+                getattr(self.model.fs[i], power_output_name)
+            )
+
+        return
+
+    def formulate_bidding_problem(self):
+
+        """
+        Formulate the bidding optimization problem by adding necessary
+        parameters, constraints, and objective function.
+
+        Arguments:
+            None
+
+        Returns:
+            None
+        """
+
+        self._add_bidding_params()
+        self._add_bidding_constraints()
+        self._add_bidding_objective()
+
+        return
+
+    def _add_bidding_params(self):
+
+        """
+        Add necessary bidding parameters to the model, i.e., market energy price.
+
+        Arguments:
+            None
+
+        Returns:
+            None
+        """
+
+        for i in self.model.SCENARIOS:
+            time_index = self.model.fs[i].power_output_ref.index_set()
+            self.model.fs[i].energy_price = pyo.Param(
+                time_index, initialize=0, mutable=True
+            )
+        return
+
+    def _add_bidding_constraints(self):
+
+        """
+        Add bidding constraints to the model, i.e., the bid curves need to be
+        nondecreasing.
+
+        Arguments:
+            None
+
+        Returns:
+            None
+        """
+
+        # declare a constraint list
+        self.model.bidding_constraints = pyo.ConstraintList()
+
+        # generate scenarios combinations
+        scenario_comb = list(combinations(self.model.SCENARIOS, 2))
+
+        for k in scenario_comb:
+            time_index = self.model.fs[k[0]].power_output_ref.index_set()
+            for t in time_index:
+                self.model.bidding_constraints.add(
+                    (
+                        self.model.fs[k[0]].power_output_ref[t]
+                        - self.model.fs[k[1]].power_output_ref[t]
+                    )
+                    * (
+                        self.model.fs[k[0]].energy_price[t]
+                        - self.model.fs[k[1]].energy_price[t]
+                    )
+                    >= 0
+                )
```

Comment:
Any reason to use a `ConstraintList` here as opposed to building a normal indexed constraint? We don't seem to mutate them.

If this is a matter of the `time_index` depending on `k[0]`, that could be handled other ways, e.g.,
```python
self.model.bidding_constraints = pyo.Constraint(self.model.SCENARIOS, self.model.SCENARIOS, pyo.Any)
```
or, keeping the `scenario_comb`:
```python
self.model.bidding_constraints = pyo.Constraint(scenario_comb, pyo.Any)
```
though obviously it would be preferable to use an explicit time index for the last component in either case. 
Note that you can simply skip the case where the scenarios are the same using the first version.